### PR TITLE
[MPM] Update MPM Elements to fix test

### PR DESCRIPTION
**Pull request by tteschemacher**
Dear all,

this is the proposed way to adapt the MPM elements to the recent changes of the GeometryDataContainer.

It hasn't reached its final stage but please feel free to give your comments.

### CODE REVIEW

**Comment by peterjwilson** [comment]
Relevant to [this](https://github.com/KratosMultiphysics/Kratos/pull/6490/commits/39701f05039b46f2dc1d9191776cb9491b43b88f) commit 
Two tests in the mpm search utility appeared incorrect. (\tests\test_search_mpm_particle.py)

### Test 1
Part 1 of test_SearchMPMParticleTriangle3DUnstructured tests that the following point (1.31967,1.85246,1.0) Is in the following tet element:
Element 0
         n0 = [0.9, 2.9, 0]
         n1 = [0, 0, 0]
         n2 = [2.9, 2.4, 0]
         n3 = [0.5, 0.5, 1.5]

The material point is point number 7 in the picture below, which is outside the element. Point 8 is the new material point location (1.31967,1.85246,0.1) which is inside the element 
image


### Test 2
Part 1 of test_SearchMPMParticleTriangle3DUnstructuredFine tests that the following point (1.31967e-07,1.85246e-07,1e-07) Is in the following tet element:
Element 0
         n0 = [9e-08, 2.9e-07, 0]
         n1 = [0, 0, 0]
         n2 = [2.9e-07, 2.4e-07, 0]
         n3 = [5e-08, 5e-08, 1.5e-07]


The material point is point number 5 in the picture below, which is outside the element. Point 6 is the new material point location (1.31967e-07,1.85246e-07,1e-08) which is inside the element 

image

File: applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_base_condition.cpp
Comment (VeronikaSinger) [inline applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_base_condition.cpp:400]: Do we need this distinction between MP_COORD and MPC_COORD as for conditions we exclusively use MPC_COORD and elements MP_COORD ?

File: applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_dirichlet_condition.h
Comment (VeronikaSinger) [inline applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_dirichlet_condition.h]: I think for the default value 0.0 would be good, as it is the same in "apply_mpm_particle_dirichlet_condition_process.py"

File: applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_base_condition.cpp
Comment (tteschemacher) [reply applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_base_condition.cpp:400]: We were actually thinking of neglecting the distinction completely. It is just introducing bugs and there is no need to do it. But open for discussion?!

File: applications/ParticleMechanicsApplication/custom_utilities/mpm_particle_generator_utility.cpp
Comment (bodhinandach) [inline applications/ParticleMechanicsApplication/custom_utilities/mpm_particle_generator_utility.cpp]: hmmm, why we dont need the id?

File: kratos/python/add_mesh_to_python.cpp
Comment (bodhinandach) [inline kratos/python/add_mesh_to_python.cpp]: This function is in the core. Is this okay for the other applications? @philbucher

File: applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_dirichlet_condition.h
Comment (tteschemacher) [reply applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_dirichlet_condition.h]: I agree, however, it is anyways an error if the default config is used.

File: applications/ParticleMechanicsApplication/python_scripts/apply_mpm_slip_boundary_process.py
Comment (bodhinandach) [inline applications/ParticleMechanicsApplication/python_scripts/apply_mpm_slip_boundary_process.py:31]: @VeronikaSinger Could you please check with the SLIP BCs? I think we still need it right?

**Review by bodhinandach — CHANGES_REQUESTED** [review]
Some comments. I will give another check as this PR is completed. Thanks @tteschemacher @PeterJosephWilson

File: applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_dirichlet_condition.h
Comment (bodhinandach) [reply applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_dirichlet_condition.h]: I think zero is better, more self explanatory, isn't it?

File: applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_base_condition.cpp
Comment (bodhinandach) [reply applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_base_condition.cpp:400]: So, initially, we intend to distinguish all the variables into two sets: MP_VAR and MPC_VAR for material point and boundary points. This is useful if you want to distinguish apply BC for both particle and boundary particles at the same time. So for displacement, velocity, or traction forces, I think it will be better to separate. However, for coordinates, we can generalize them if you like.

File: applications/ParticleMechanicsApplication/custom_utilities/mpm_particle_generator_utility.cpp
Comment (tteschemacher) [reply applications/ParticleMechanicsApplication/custom_utilities/mpm_particle_generator_utility.cpp]: Through the CalculateOnIntegrationPoints we can directly access it. Thus, no need to duplicate it.

File: kratos/python/add_mesh_to_python.cpp
Comment (tteschemacher) [reply kratos/python/add_mesh_to_python.cpp]: This is a separate PR. #6499 .

File: applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_base_condition.cpp
Comment (tteschemacher) [reply applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_base_condition.cpp:400]: You can distinguish already with condition or element.

However, it is discussion, but as our elements are very complex we should consider neglecting using both (for maintenance reasons - and no real disadvantage). Having one Base class will ease many procedures.

File: applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_dirichlet_condition.h
Comment (tteschemacher) [reply applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_dirichlet_condition.h]: still, it is an wrong if it is not applied.

File: applications/ParticleMechanicsApplication/custom_utilities/mpm_particle_generator_utility.cpp
Comment (bodhinandach) [reply applications/ParticleMechanicsApplication/custom_utilities/mpm_particle_generator_utility.cpp]: @tteschemacher Can you check the coupling test too implemented in CoSimulation Application? They are related to accessing the id if I am not mistaken.

File: applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_coupling_interface_condition.cpp
Comment (bodhinandach) [inline applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_coupling_interface_condition.cpp:214]: is m_unit_normal normalized? Can you always Normalize to make sure? This is essentially needed if we do a coupling with FEM, cause FEM will only send the direction.

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.cpp
Comment (bodhinandach) [inline applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.cpp:1750]: I am not sure whether this check is necessary.

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.hpp
Comment (bodhinandach) [inline applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.hpp:86]: I just want to confirm: are these variables created because the original variable registered under SetValue and GetValue are removed?

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.hpp
Comment (bodhinandach) [inline applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.hpp:135]: This will limit you to access the third direction strain and stress under plane stress and plane strain assumption. For certain material models like CamClay, you might need the third directions component even in 2D.

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_quadrilateral.hpp
Comment (bodhinandach) [inline applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian_quadrilateral.hpp:136]: idem

**Review by bodhinandach — CHANGES_REQUESTED** [review]
@tteschemacher @peterjwilson Can we discuss this PR in a skype call next week?

File: applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_coupling_interface_condition.cpp
Comment (peterjwilson) [reply applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_coupling_interface_condition.cpp:214]: Yep its normalized on input.
https://github.com/KratosMultiphysics/Kratos/blob/2cc6f47d6bbc432f007d7816e3e77642279c31f7/applications/ParticleMechanicsApplication/custom_conditions/particle_based_conditions/mpm_particle_penalty_dirichlet_condition.cpp#L314

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.cpp
Comment (peterjwilson) [reply applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.cpp:1750]: The std vector wrapper should always have size one since its only a wrapper of one piece of information. Its added safety to include the check - it enforces that the code is used how it was intended to be.

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.hpp
Comment (peterjwilson) [reply applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.hpp:86]: Yes that's right, they are created on the local element now instead of the geometry due to changes in GeometryDataContainer - this is the core of this PR.

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.hpp
Comment (peterjwilson) [reply applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.hpp:135]: Thanks for the heads up. From your experience, how was this handled before? The element must have known the correct strain size somehow?

**Comment by peterjwilson** [comment]
> @tteschemacher @peterjwilson Can we discuss this PR in a skype call next week?

@bodhinandach Working through your recent comments, it seems far quicker and more precise to work through each concrete issue on Github. Thanks for raising them.

@bodhinandach Did you verify if the co-sim test with FEM is working? I believe this check, and the strain size issues are the only blocks to your approval remaining?

**Comment by peterjwilson** [comment]
Ping @bodhinandach and @VeronikaSinger . It would be nice to reach resolution on this and either get approvals or concrete specific change requests from you guys.

File: .github/workflows/configure.sh
Comment (philbucher) [inline .github/workflows/configure.sh:42]: to me this makes sense

How about adding it also to the [Win-CI](https://github.com/KratosMultiphysics/Kratos/blob/master/.github/workflows/ci.yml#L103-L108)?

furthermore you could also add it to the [nightly build](https://github.com/KratosMultiphysics/Kratos/blob/master/.github/workflows/nightly_build.yml#L81-L89)

can you please also check the test in the CoSimApp?

File: applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.hpp
Comment (bodhinandach) [reply applications/ParticleMechanicsApplication/custom_elements/updated_lagrangian.hpp:135]: I checked it, it should be fine.

**Comment by bodhinandach** [comment]
@peterjwilson I am fine. But we need to check the one in CoSimApp, should I do it?

**Comment by philbucher** [comment]
you can check the nightly build (which is also triggered by this PR) and see that the CoSim Test fails. This failure is related to #6492 and can be easily be fixed, see #6523

**Comment by peterjwilson** [comment]
> @peterjwilson I am fine. But we need to check the one in CoSimApp, should I do it?

Nah I'll check it out, thanks for offering though

File: applications/ParticleMechanicsApplication/python_scripts/apply_mpm_slip_boundary_process.py
Comment (VeronikaSinger) [reply applications/ParticleMechanicsApplication/python_scripts/apply_mpm_slip_boundary_process.py:31]: it is still needed for the check in mpm_boundary_rotation_utility.h whether is_penalty is true or false

**Comment by tteschemacher** [comment]
@philbucher it seems that the remaining errors are not coming from the MPM. Could you please have a check, here?

**Comment by tteschemacher** [comment]
@bodhinandach could you remove your change request if you do not have any further requests? Thanks!

**Comment by tteschemacher** [comment]
This PR passes all required tests. The tests which are failing in the nightly tests are not related to the mpm.
@VeronikaSinger @bodhinandach @RiccardoRossi @antonialarese any further suggestions?

**Comment by bodhinandach** [comment]
@tteschemacher can you give me time, I will check it tonight with the CoSim. Sorry for keeping you waiting.

**Comment by tteschemacher** [comment]
@bodhinandach looking forward to read your review tomorrow morning! ;)

File: applications/ParticleMechanicsApplication/python_scripts/apply_mpm_slip_boundary_process.py
Comment (bodhinandach) [reply applications/ParticleMechanicsApplication/python_scripts/apply_mpm_slip_boundary_process.py:31]: @VeronikaSinger Can you please check this? I think this is still needed

**Review by bodhinandach — COMMENTED** [review]
I am okay. Test in CoSim is now running. @VeronikaSinger, can you just confirm whether the IS_STRUCTURE is still needed. I would say that it's needed.

**Comment by tteschemacher** [comment]
@VeronikaSinger tested some examples and it seems to work properly.

@RiccardoRossi @antonialarese any further suggestions?

**Comment by tteschemacher** [comment]
@RiccardoRossi @antonialarese @bodhinandach @VeronikaSinger any further suggestions?

**Comment by philbucher** [comment]
I have another comment that I will post soon

**Comment by philbucher** [comment]
> I have another comment that I will post soon

Nevermind, you already fixed it ;)

**Comment by tteschemacher** [comment]
Anyways thanks @philbucher .

I don't think that we'll be able to fix everything. However, small changes will easy within new PRs.

**Comment by philbucher** [comment]
from my side this is PR is ok

@KratosMultiphysics/technical-committee just for the record this PR adds the MPM-App to the CI.

**Comment by tteschemacher** [comment]
That means, thanks to code owner... 

We need your approval @RiccardoRossi !!

**Comment by philbucher** [comment]
> That means, thanks to code owner...

no, this is still not enforced. Even if it will always be teams not individual ppl

@bodhinandach is blocking, not @RiccardoRossi (this time :D)

**Review by bodhinandach — APPROVED** [review]
I am good with the PR! Sorry for the block. I am at the moment asking @VeronikaSinger to take a look at it one more time, once she approves, feel free to merge :) Thanks @tteschemacher @PeterJosephWilson

**Comment by VeronikaSinger** [comment]
I just want to check another example as I am still wondering if we need the "IS_STRUCTURE" variable

**Comment by VeronikaSinger** [comment]
Ok thanks I am fine.

**Review by VeronikaSinger — APPROVED** [review]

**Comment by tteschemacher** [comment]
Thanks all for the contribution in the discussion.

Thank you @PeterJosephWilson for fixing all the tests!

I am merging now, that no one can change it's opinion, anymore. If you'll find more bugs you can keep them ;)

**Comment by antonialarese** [comment]
Thanks @tteschemacher @PeterJosephWilson  and all the team!!! Great work!

**Pull request merged**
